### PR TITLE
Ensure file is closed promptly in case of error

### LINKED ISSUE #35566: BUG: corrupted SAS datasets retain file handles
- [ ] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas.

- [X] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
from pathlib import Path

path = ...
p = Path(path, 'corrupt.sas7bdat')
try:
    pd.read_sas(p)
except:
    input("Press enter to release file")
```

#### Problem description

When I run the above script with a corrupt database like the one attached [corrupt.zip](https://github.com/pandas-dev/pandas/files/5028407/corrupt.zip) the Python process my Windows system ends up holding on to the open file.

I can tell that the file has not been correctly closed on error since, for example, other processes cannot open the same file.

#### Expected Output

While the main Python process is waiting for input, other processes should be able to read the given file.

I can't find any way of finding out if the current Python process has an open file handle on a given file, but it is expected behaviour that other processes should be able to access the file before

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : d9fff2792bf16178d4e450fe7384244e50635733
python           : 3.7.1.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.17763
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 9, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None
pandas           : 1.1.0
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 10.0.1
setuptools       : 39.0.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
numba            : None

</details>

### PR DESCRIPTION
**Pull request by rxxg**
Fixes #35566

### CODE REVIEW

**Comment by jbrockmendel** [comment]
can you add a test

**Comment by rxxg** [comment]
Very difficult to add a test since the failure is during the constructor meaning that we have no references to the failed state.
I have pulled something together, but it requires a dependency on psutils and assumes that the test file is on a local disk.  Is that acceptable?

File: pandas/io/sas/sasreader.py [hunk 1]
```diff
@@ -128,6 +128,7 @@ def read_sas(
     if iterator or chunksize:
         return reader
```
Comment (topper-123) [inline pandas/io/sas/sasreader.py:129]: The problem is potentialy here also, e.g. if the user forgets to wrap his code in a try/finally. 

Could it be an idea to add ``__enter__`` & ``__exit__``  to ReaderBase and rearrange to use ``with`` blocks, both here and below?

**Comment by topper-123** [comment]
A dependency on external libs for this is not great IMO. Are there tests for closed files for other file formats, e.g. excel or csv files? Could you try to emulate them?

File: pandas/io/sas/sasreader.py (hunk 1 above)
Comment (rxxg) [reply pandas/io/sas/sasreader.py:129]: What the user does with the Reader object is not pandas' concern ;)
The important thing is not to leave any created objects in an inconsistent state.

**Comment by rxxg** [comment]
OK, let's follow up in the new PR #35587 (sorry, I deleted my repo in the meantime, so need a new PR).

**Pull request closed without merging**